### PR TITLE
Fix CohereAsr training-loss double-shift (train against labels, not labels[..., 1:])

### LINKED ISSUE #46894: CohereASR training-loss double-shift bug (same pattern as Moonshine #46784)
### System Info

- `transformers` version: 5.13.0.dev0 (main branch at commit dc5a497c18)
- - Platform: Windows
- - - Python version: 3.x
- - - - Using GPU: N/A (code-level bug, not runtime)

### Who can help?

@eustlb @ebezzam @vasqu

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

aIn `src/transformers/models/cohere_asr/modular_cohere_asr.py`, `CohereAsrForConditionalGeneration.forward()` has a double-shift bug:

1. Labels are shifted right via `shift_tokens_right()` (line 483) to create `decoder_input_ids`
2. 2. Then `self.loss_function()` is called (line 503), which maps to `ForCausalLMLoss`
3. 3. `ForCausalLMLoss` shifts labels again internally (`labels[..., 1:]` in `loss/loss_utils.py` line 63-64)
This means the model trains against `labels[..., 1:]` instead of `labels`.

This is the **exact same bug** that was fixed in Moonshine in PR #46784 (commit d8c235494e). The Moonshine fix replaced `self.loss_function()` with `CrossEntropyLoss()`, but `CohereAsr` — which inherits from Moonshine and overrides `forward` — was not updated.

```python
# Buggy code (line 503 in modular_cohere_asr.py):
loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size)

# Should be:
loss_fct = CrossEntropyLoss()
loss = loss_fct(logits.reshape(-1, self.config.vocab_size), labels.reshape(-1))
```

### Expected behavior

Training loss should be computed directly against `labels` without any additional shifting, since `shift_tokens_right()` already handles the decoder input alignment. This matches the behavior of Whisper, Bart, and the now-fixed Moonshine model. The fix should use `CrossEntropyLoss()` instead of `self.loss_function()`, as done in Moonshine PR #46784.

### PR DESCRIPTION
**Pull request by eliaghazal**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47090)

# What does this PR do?

`CohereAsrForConditionalGeneration.forward` right-shifts `labels` into `decoder_input_ids` via `shift_tokens_right`, then computes the loss with `self.loss_function` (`ForCausalLMLoss`), which shifts the labels a second time internally. The model therefore trains against `labels[..., 1:]` instead of `labels`.

Verified on main (`b70d02fc72`) with a tiny random-init model:

```
model loss:            4.990426
aligned CE (correct):  4.985212
double-shift CE (bug): 4.990426
```

This is the same bug fixed for Moonshine in #46784. CohereAsr inherits from Moonshine but overrides `forward`, so that fix didn't propagate.

The fix keeps `self.loss_function` and passes the (already aligned) labels as `shift_labels`, so `ForCausalLMLoss` skips its internal shift while preserving the `num_items_in_batch` handling needed for correct loss reduction under gradient accumulation — same pattern as Florence2 after #46898. Changed in `modular_cohere_asr.py` and regenerated `modeling_cohere_asr.py` with the modular converter.

Also adds the same regression test as Moonshine's (`test_training_loss_no_double_shift`, with and without `-100` padding), extended with an assertion for the `num_items_in_batch` (summed) reduction path. It fails on current main and passes with this fix; the full `tests/models/cohere_asr` file passes (124 passed, 138 skipped).

Fixes #46894
Closes #46895, closes #46958 (earlier PRs for the same issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the forum? Please add a link to it if that's the case. → #46894
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

@eustlb @vasqu

### CODE REVIEW

**Review by eustlb — COMMENTED** [review]
Nice catch! See the comment (84e2058) I added to make this explicit, but the issue here is that we were mixing two different conventions:

- whisper/ legacy encoder-decoder convention: we pass labels, and forward builds decoder_input_ids from them by right-shifting: `[the, black, cat]` → `[bos, the, black, cat]`
- llama-like / causal LM convention: `labels == input_ids`. `input_ids` and `labels` are `[bos, the, black, cat]` and `ForCausalLMLoss` shifts labels into `[the, black, cat]`, ie we do `[bos, the, black, cat]` →  `[the, black, cat]`

Ideally we would modernize Whisper and similar models to follow the causal LM convention, but that would require too much BC handling, so this fix LGTM 👌

**Review by eustlb — APPROVED** [review]

File: src/transformers/models/cohere_asr/modeling_cohere_asr.py [hunk 1]
```diff
@@ -637,7 +638,8 @@ def forward(
 
         loss = None
         if labels is not None:
-            loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.vocab_size)
+            loss_fct = CrossEntropyLoss()
+            loss = loss_fct(logits.reshape(-1, self.config.vocab_size), labels.reshape(-1))
```
Comment (vasqu) [inline src/transformers/models/cohere_asr/modeling_cohere_asr.py]: My only concern with this is that we don't really support proper gradient accumulation with this anymore which is the reason behind the whole loss function tbh --> gradient accumulation sums and divides only at the end (before it divided on each step)

**Review by vasqu — COMMENTED** [review]
So we don't want to use the loss function in any case? Note that this happens on several models, not only cohere asr - would be nice to collectively fix them in one PR?

Also cc @ebezzam because we talked about it

**Comment by vasqu** [comment]
Can we add closes #... for the other PRs so we focus on only this one?

File: src/transformers/models/cohere_asr/modeling_cohere_asr.py (hunk 1 above)
Comment (eliaghazal) [reply src/transformers/models/cohere_asr/modeling_cohere_asr.py]: Fixed in 4629fa3 — kept `self.loss_function` and passed the already-aligned labels as `shift_labels`, so the internal shift is skipped but the `num_items_in_batch` reduction still applies (same pattern as Florence2 after #46898). Also extended the regression test with an assertion for the `num_items_in_batch` path.

**Comment by eliaghazal** [comment]
@vasqu good catch on the gradient accumulation — dropping `self.loss_function` did lose the `num_items_in_batch` handling. I've updated the PR (4629fa3) to keep `self.loss_function` and pass the labels as `shift_labels` instead: forward already right-shifts them into `decoder_input_ids`, so they're aligned with the logits and `ForCausalLMLoss` can skip its internal shift while still summing and dividing by `num_items_in_batch`. This is the same pattern Florence2 uses since #46898. The regression test now also covers the `num_items_in_batch` reduction path.

On fixing collectively — from a sweep of models mixing the whisper-style label shift with `ForCausalLMLoss`:

- **Florence2** — already fixed with the `shift_labels` pattern (#46898)
- **Moonshine** — fixed in #46784, but with a plain `CrossEntropyLoss`, so it has the same grad-accum gap you raised here; happy to open a follow-up PR moving it to the `shift_labels` pattern
- **PPFormulaNet** — builds `decoder_input_ids` by shifting `input_ids` while the loss shifts `labels` internally, so it looks affected too (depends on the intended labels convention); can check it in the same follow-up

I'd keep this PR scoped to CohereAsr and do the Moonshine/PPFormulaNet pass in a follow-up, unless you'd prefer it all in one.

Also added the closes-references for the earlier duplicate PRs (#46895, #46958) to the description.

File: src/transformers/models/cohere_asr/modular_cohere_asr.py
Comment (vasqu) [inline src/transformers/models/cohere_asr/modular_cohere_asr.py:1]: Welp now we are getting close to #46895 😅 cc @SunMarc @eustlb to coordinate a bit

**Comment by eliaghazal** [comment]
Fair point 😅 — reading through #46895, the review there converged on this same `shift_labels` approach before my update did, so credit to @sharmax-vikas. No attachment to which PR lands — happy to close this one in favor of #46895 if that's the cleaner path.

Whichever one you pick, two bits from here may be worth carrying over: the regression test also asserts the `num_items_in_batch` (summed) reduction path, and the change is made in `modular_cohere_asr.py` with `modeling_cohere_asr.py` regenerated by the converter.

Either way I'll follow up on the Moonshine/PPFormulaNet sweep discussed above once this settles.

**Comment by vasqu** [comment]
Yep, not on you. Imo it's just about coordination now and imo probably best to sweep all affected models at once

**Comment by eliaghazal** [comment]
Happy to do the sweep in this PR then. I'd extend it to move Moonshine off the plain CrossEntropyLoss (it has the same num_items_in_batch gap) and check/fix PPFormulaNet, each with the same regression test including the grad accumulation reduction path. If you'd rather do it in #46895 with @sharmax-vikas driving that's fine too, just say the word and I'll share what I have.

**Comment by vasqu** [comment]
Let's wait on @SunMarc and @eustlb first, also happy with either :hugs: expecting sometime next week to handle this properly then.

**Comment by SunMarc** [comment]
Then maybe we can merge this https://github.com/huggingface/transformers/pull/46895 and @eliaghazal, if you are willing, maybe you can propagate this change to other models so that the have the same behavior across all models that have this isssue ?

**Comment by eliaghazal** [comment]
@SunMarc done — this PR now carries the propagation:

- **Moonshine**: #46784 fixed the double-shift but left a plain `CrossEntropyLoss`, which drops `num_items_in_batch` — so loss reduction under gradient accumulation is wrong. Moved it to `self.loss_function(..., shift_labels=labels)` like the rest.
- **PPFormulaNet**: inherits Florence2 but overrides `forward`, so it missed #46898 and still double-shifts (also reported in #47317). Same `shift_labels` fix.
- Regression tests for both exercise the grad-accumulation path with a global token count that differs from the per-step count, so the summed path is distinguishable from a plain mean.
- Audited the rest of the codebase for models that right-shift labels into `decoder_input_ids` and call `self.loss_function` without `shift_labels`: the only other hit is T5Gemma, which is fine — it uses `loss_type="ForMaskedLM"`, which doesn't shift.

Coordination so we don't trip over each other: if #46895 lands first I'll rebase and drop the CohereASR commits here; and #46903 (@OmkumarSolanki) also covers PPFormulaNet with a config-gated common test — happy to drop my PPFormulaNet half in deference, whichever way you prefer to slice it.

**Comment by eliaghazal** [comment]
Closing this one — every fix it carries has landed upstream through #46895 (CohereASR) and #46903 (Moonshine, MoonshineStreaming, PPFormulaNet), both merged. Confirmed on current `main`: all four files already carry the `shift_labels` fix.

Thanks @SunMarc and @vasqu for the review time, and nice work @sharmax-vikas / @OmkumarSolanki on getting the equivalent fixes in first.
